File: src/main/java/com/oycm/algorithm/f/basic/Solution_8.scala
package com.oycm.algorithm.f.basic

object Solution_8 {

  /**
   * 3361. 两个字符串的切换距离 1533
   * https://leetcode.cn/problems/shift-distance-between-two-strings/description/
   *
   * 一次操作中，可以选择 s 中的一个下标执行以下操作之一：
   *  - 将 s[i] 切换为字母表中的下一个字母, s[i] = 'z' 切换 成 s[i] = 'a' 切换的代价 nextCost[25]
   *  - 将 s[i] 切换为字母表中的上一个字母, s[i] = 'a' 切换 成 s[i] = 'z' 切换的代价 previousCost[0]
   *
   * 切换距离：s 切换成 t 的最少操作代价之和
   *
   * @param s            长为 n 的字符串
   * @param t            长为 n 的字符串
   * @param nextCost     长 26 的数组，切换成下一个字符的花费
   * @param previousCost 长 26 的数组，切换成上一个字符的花费
   * @return
   */
  def shiftDistance(s: String, t: String, nextCost: Array[Int], previousCost: Array[Int]): Long = {
    /*
    切换的过程中，只能沿一个方向，要么是顺切换，要么是逆切换，才能最小操作代价。除非都是 0 的情况下，可以随便来回切
    当字符 i < j
      nextCost: nextCost(i) + nextCost(i+1) + nextCost(j-1)
        nextSums(j) - nextSums(i)
      previousCost: previousCost(i) + previousCost(i-1) + previousCost(0) + previousCost(25) + .. + previousCost(j+1)
        preSums(26) - (preSums(j+1) - preSums(i+1))
     当字符 i > j
      nextCost: nextCost(i) + nextCost(i+1) + ... + nextCost(25) + nextCost(0) + nextCost(j-1)
        nextSums(26) - (nextSums(i) - nextSums(j))
      previousCost: previousCost(i) + previousCost(i-1) + ... + previousCost(j+1)
        preSums(i+1) - preSums(j+1)
    */
    var ans = 0L
    // 注意前缀和溢出问题
    val nextSums = Array.fill(27)(0L)
    val preSums = Array.fill(27)(0L)
    for(i <- nextCost.indices) {
      nextSums(i + 1) = nextSums(i) + nextCost(i)
      preSums(i + 1) = preSums(i) + previousCost(i)
    }
    for (k <- s.indices) {
      val i = s(k) - 'a'
      val j = t(k) - 'a'
      if (i < j) {
        ans += Math.min(nextSums(j) - nextSums(i), preSums(26) - (preSums(j+1) - preSums(i+1)))
      } else if (i > j) {
        ans += Math.min(nextSums(26) - (nextSums(i) - nextSums(j)), preSums(i+1) - preSums(j+1))
      }
    }
    ans
  }

  def main(args: Array[String]): Unit = {

  }

}

File: src/main/java/com/oycm/algorithm/bb/Solution_8.java
package com.oycm.algorithm.bb;

import java.util.Arrays;

public class Solution_8 {

    /**
     * 1385. <a href="https://leetcode.cn/problems/find-the-distance-value-between-two-arrays/description/">两个数组间的距离值</a> 1235
     *
     * @param arr1
     * @param arr2
     * @param d
     * @return
     */
    public int findTheDistanceValue(int[] arr1, int[] arr2, int d) {
        /*
        距离值定义：为符合此距离要求的元素数目。
        对于元素 arr1[i] ，不存在任何元素 arr2[j] 满足 |arr1[i]-arr2[j]| <= d。
         */
        /*
        由于对 arr1, arr2 下标没有要求，可以先对数组进行排序。
        arr1[i], arr2 不存在元素在 [arr1[i] - d, arr1[i] + d]
        题解思路：排序后维持每个 arr1[i] - d >= arr2[j] 的最小下标 j
        如果 j == arr2.length 较小值都满足 arr2[j] < arr1[i] - d, 数轴的角度来看，肯定后续的符合要求
        如果 arr2[j] <= x + d, 则不符合要求，在区间范围
         */
        int ans = 0;
        Arrays.sort(arr1);
        Arrays.sort(arr2);
        int j = 0;
        for (int x : arr1) {
            while (j < arr2.length && arr2[j] < x - d) {
                j++;
            }
            if (j == arr2.length || arr2[j] > x + d) {
                ans++;
            }
        }

        return ans;
    }

}
